File: src/bot.js
const { Client, GatewayIntentBits, Events, Collection, ChannelType, Routes } = require('discord.js');
const { OpenAI } = require('openai');
const dotenv = require('dotenv');
const ConversationHandler = require('./utils/ConversationHandler');
const VoiceHandler = require('./utils/VoiceHandler');
const { loadCommands } = require('./utils/commandLoader');
const { cleanup, restoreVoiceState } = require('./cleanup');

// Load environment variables
dotenv.config();

// Create Discord client with necessary intents
const client = new Client({
    intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
        GatewayIntentBits.GuildVoiceStates,
        GatewayIntentBits.DirectMessages,
        GatewayIntentBits.DirectMessageTyping,
        GatewayIntentBits.DirectMessageReactions,
        GatewayIntentBits.GuildMembers,
        GatewayIntentBits.GuildPresences,
        GatewayIntentBits.MessageContent,  // Required for message content in DMs
        GatewayIntentBits.GuildMessageTyping,
        GatewayIntentBits.GuildMessageReactions,
        32767  // All intents (temporary for testing)
    ],
    partials: [
        'CHANNEL',      // Required to receive DMs
        'MESSAGE',      // Required to receive messages
        'USER',         // Required for user partial info
        'GUILD_MEMBER', // Required for member info
        'REACTION',     // Required for message reactions
        'DM'           // Required specifically for DM channels
    ]
});

// Initialize OpenAI client
const ai = new OpenAI({
    apiKey: process.env.DEEPSEEK_API_KEY,
    baseURL: process.env.DEEPSEEK_API_BASE
});

// Initialize voice handler
const voiceHandler = new VoiceHandler(client);
client.voiceHandler = voiceHandler;  // Attach to client

// Initialize conversation handler with voice capabilities
client.conversationHandler = new ConversationHandler(20, voiceHandler);

// Initialize commands collection
client.commands = new Collection();

// Load commands
loadCommands(client);

// Add ready event handler to log when bot is ready
client.once(Events.ClientReady, () => {
    console.log(`Logged in as ${client.user.tag}!`);
    console.log('Bot is ready to receive messages!');
});

// Generate AI response for text chat
async function generateTextResponse(channelId, newMessage, isDM = false, userId = null) {
    console.log(`\n=== Generating ${isDM ? 'DM' : 'text'} response for ${isDM ? `user ${userId}` : `channel ${channelId}`} ===`);
    
    // Get conversation history
    const history = isDM 
        ? client.conversationHandler.getDMHistory(userId)
        : client.conversationHandler.getTextHistory(channelId);
    
    // Keep system message plus last N messages
    const messages = [
        history[0], // System message
        ...history.slice(Math.max(1, history.length - client.conversationHandler.maxHistory + 1)),
        newMessage
    ];
    
    // Log the context window
    console.log(`\n=== ${isDM ? 'DM' : 'Text'} Chat Context Window ===`);
    messages.forEach((msg, index) => {
        if (index === 0) {
            console.log('\n[System Message]');
            console.log(msg.content.split('\n').map(line => `  ${line}`).join('\n'));
        } else {
            console.log(`\n[${msg.role.charAt(0).toUpperCase() + msg.role.slice(1)} Message]`);
            console.log(`  ${msg.content}`);
        }
    });
    console.log('\n=== End Context Window ===\n');
    
    try {
        console.log('Sending request to DeepSeek API...');
        const response = await ai.chat.completions.create({
            model: "deepseek-chat",
            messages: messages,
            max_tokens: 1000,
            temperature: 0.7,
            user: newMessage.name
        });
        
        const responseText = response.choices[0].message.content;
        console.log('Received text response:', responseText);
        return responseText;
    } catch (error) {
        console.error("Error generating text response:", error);
        throw error;
    }
}

// Attach generateTextResponse to client so commands can use it
client.generateTextResponse = generateTextResponse;

// Message event handler
client.on(Events.MessageCreate, async message => {
    try {
        // Fetch the channel if it's partial
        if (message.channel?.partial) {
            await message.channel.fetch();
        }

        const isDM = message.channel.type === 1;

        // Ignore messages from bots (including our own messages)
        if (message.author.bot || (isDM && message.author.id === client.user.id)) return;
        
        console.log(`Received ${isDM ? 'DM' : 'server'} message:`, {
            content: message.content,
            channelType: message.channel.type,
            isDM: isDM,
            author: message.author.tag,
            guildId: message.guild?.id || 'None (DM)'
        });
        
        // Process commands if message starts with prefix
        if (message.content.startsWith('!')) {
            const args = message.content.slice(1).trim().split(/ +/);
            const commandName = args.shift().toLowerCase();
            
            const command = client.commands.get(commandName);
            if (!command) return;
            
            try {
                await command.execute(message, args);
            } catch (error) {
                console.error(error);
                await message.reply('There was an error executing that command!');
            }
            return;
        }
        
        // Format the new message
        const newMessage = client.conversationHandler.formatUserMessage(message.author.id, message.content);
        
        // Generate response if message is in DM or bot is mentioned in server
        if (isDM || (!isDM && message.mentions.has(client.user))) {
            console.log('Preparing to respond to message...');
            await message.channel.sendTyping();
            
            try {
                // Add the user's message to appropriate history
                if (isDM) {
                    client.conversationHandler.addDMMessage(message.author.id, newMessage);
                } else {
                    client.conversationHandler.addTextMessage(message.channel.id, newMessage);
                }
                
                // Generate and get the response
                const responseText = await generateTextResponse(
                    message.channel.id,
                    newMessage,
                    isDM,
                    isDM ? message.author.id : null
                );
                
                // Add bot's response to appropriate history
                const assistantMessage = client.conversationHandler.formatAssistantMessage(responseText);
                if (isDM) {
                    client.conversationHandler.addDMMessage(message.author.id, assistantMessage);
                } else {
                    client.conversationHandler.addTextMessage(message.channel.id, assistantMessage);
                }
                
                // Split and send response if too long
                const maxLength = 1900;
                for (let i = 0; i < responseText.length; i += maxLength) {
                    const chunk = responseText.substring(i, i + maxLength);
                    await message.channel.send(chunk);
                }
            } catch (error) {
                console.error('Error processing message:', error);
                await message.reply(`Sorry <@${message.author.id}>, I encountered an error: ${error.message}`);
            }
        } else if (!isDM) {
            // If not generating a response and not in DM, just add the message to text history
            client.conversationHandler.addTextMessage(message.channel.id, newMessage);
        }
    } catch (error) {
        console.error('Error in message event handler:', error);
    }
});

// Debug event to check what events we're receiving
client.on('raw', packet => {
    if (packet.t === 'MESSAGE_CREATE') {
        const data = packet.d;
        console.log('Raw MESSAGE_CREATE event:', {
            channelID: data.channel_id,
            type: data.channel_type,
            content: data.content,
            author: data.author,
            timestamp: new Date().toISOString()
        });
        
        // Try to manually emit the message create event
        if (data.channel_type === 1) { // If it's a DM
            const channel = client.channels.cache.get(data.channel_id) || {
                id: data.channel_id,
                type: 1,
                partial: false,
                send: async (content) => {
                    const payload = typeof content === 'string' ? { content } : content;
                    return client.rest.post(Routes.channelMessages(data.channel_id), { body: payload });
                },
                sendTyping: async () => {
                    return client.rest.post(Routes.channelTyping(data.channel_id));
                },
                messages: new Collection(),
                client: client,
                isTextBased: () => true,
                isDMBased: () => true
            };
            
            const message = {
                id: data.id,
                content: data.content,
                author: {
                    ...data.author,
                    tag: `${data.author.username}#${data.author.discriminator}`,
                    bot: false
                },
                channel: channel,
                guild: null,
                type: data.type,
                partial: false,
                mentions: { has: () => false },
                reply: async (content) => channel.send(content)
            };
            
            console.log('Emitting constructed message event:', {
                content: message.content,
                author: message.author.tag,
                channelId: message.channel.id
            });
            
            client.emit(Events.MessageCreate, message);
        }
    }
});

// Add debug handler for all message events
client.on('debug', info => {
    console.log('Debug Info:', info);
});

// Add error handler
client.on('error', error => {
    console.error('Discord client error:', error);
});

// Set up cleanup handlers
cleanup(client);

// Log in to Discord
client.login(process.env.DISCORD_TOKEN); 

File: src/bot/bot.py
import os
import discord
from discord.ext import commands
from dotenv import load_dotenv
from openai import AsyncOpenAI
import asyncio
from typing import Optional
import logging
from aiohttp import web

from ..utils.conversation_handler import ConversationHandler
from ..voice.voice_handler import VoiceHandler

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger('Bot')

# Load environment variables
load_dotenv()
TOKEN = os.getenv('DISCORD_TOKEN')
DEEPSEEK_API_KEY = os.getenv('DEEPSEEK_API_KEY')
DEEPSEEK_API_BASE = os.getenv('DEEPSEEK_API_BASE')

# Add debug print
if not TOKEN:
    raise ValueError("No token found in environment variables! Check your .env file.")
logger.info("Token loaded: %s...", TOKEN[:10] if TOKEN else "None")

# Initialize OpenAI client for DeepSeek
ai_client = AsyncOpenAI(
    api_key=DEEPSEEK_API_KEY,
    base_url=DEEPSEEK_API_BASE
)

# Set up bot with command prefix '!'
intents = discord.Intents.default()
intents.message_content = True
intents.voice_states = True  # Required for voice functionality
bot = commands.Bot(command_prefix='!', intents=intents)

# Initialize handlers
conversation_handler = ConversationHandler()
voice_handler = VoiceHandler(bot)

# Create web app for health checks
app = web.Application()

async def health_check(request):
    """Health check endpoint"""
    return web.Response(text='OK', status=200)

app.router.add_get('/health', health_check)

async def generate_response(id: int, new_message: dict, guild_id: Optional[int] = None, is_voice: bool = False) -> str:
    """Generate a response using the AI model"""
    logger.info("Generating response for %s (voice: %s)", id, is_voice)
    
    # Get conversation history
    history = conversation_handler.get_history(id, is_voice)
    logger.info("Got conversation history with %d messages", len(history))
    
    # Keep system message plus last N messages
    messages = [
        history[0],  # System message
        *history[max(1, len(history) - conversation_handler.max_history + 1):],  # Last N-1 messages
        new_message  # Current message
    ]
    
    try:
        # Call DeepSeek API
        logger.info("Calling DeepSeek API")
        response = await ai_client.chat.completions.create(
            model="deepseek-chat",
            messages=messages,
            max_tokens=1000,
            temperature=0.7,
            user=new_message.get('name')
        )
        
        response_text = response.choices[0].message.content
        logger.info("Generated response: %s...", response_text[:50])
        return response_text
    except Exception as e:
        logger.error("Error generating response: %s", e)
        raise

async def handle_voice_message(guild_id: int, text: str) -> None:
    """Handle a voice message and generate a response"""
    logger.info("Handling voice message in guild %d: %s", guild_id, text)
    
    # Format the message
    user_id = "voice_user"  # You might want to track the actual speaking user
    message = conversation_handler.format_user_message(user_id, text)
    
    try:
        # Add message to history
        logger.info("Adding message to voice history")
        conversation_handler.add_message(guild_id, message, is_voice=True)
        
        # Generate response
        logger.info("Generating response for voice message")
        response_text = await generate_response(guild_id, message, guild_id, is_voice=True)
        
        # Add response to history
        logger.info("Adding response to voice history")
        conversation_handler.add_message(
            guild_id,
            conversation_handler.format_assistant_message(response_text),
            is_voice=True
        )
        
        # Speak the response
        logger.info("Speaking response: %s...", response_text[:50])
        await voice_handler.speak(guild_id, response_text)
    except Exception as e:
        logger.error("Error handling voice message: %s", e)

@bot.event
async def on_ready():
    logger.info('%s has connected to Discord!', bot.user)
    logger.info('Bot is in %d guilds', len(bot.guilds))

@bot.event
async def on_message(message):
    # Ignore messages from this bot
    if message.author == bot.user:
        return
        
    # Process commands first
    await bot.process_commands(message)
    
    # Handle YAGPDB.xyz welcome messages
    if message.author.id == 204255221017214977:  # YAGPDB.xyz's ID
        if any(keyword in message.content.lower() for keyword in ['welcome', 'joined', 'welcomed']):
            async with message.channel.typing():
                new_message = conversation_handler.format_user_message(
                    str(message.author.id),
                    f"I just welcomed a new user with this message: {message.content}"
                )
                
                conversation_handler.add_message(message.channel.id, new_message)
                response_text = await generate_response(message.channel.id, new_message, message.guild.id)
                conversation_handler.add_message(
                    message.channel.id,
                    conversation_handler.format_assistant_message(response_text)
                )
                
                await asyncio.sleep(1)
                await message.channel.send(response_text)
            return
    
    # If message starts with prefix, don't process it as conversation
    if message.content.startswith(bot.command_prefix):
        return
        
    # Add message to history
    new_message = conversation_handler.format_user_message(str(message.author.id), message.content)
    conversation_handler.add_message(message.channel.id, new_message)
    
    # If bot is mentioned or message is in DM, generate and send response
    if bot.user.mentioned_in(message) or isinstance(message.channel, discord.DMChannel):
        async with message.channel.typing():
            try:
                response_text = await generate_response(message.channel.id, new_message, message.guild.id)
                conversation_handler.add_message(
                    message.channel.id,
                    conversation_handler.format_assistant_message(response_text)
                )
                
                # Split and send response
                max_length = 1900
                chunks = [response_text[i:i + max_length] for i in range(0, len(response_text), max_length)]
                for chunk in chunks:
                    await message.channel.send(chunk)
                    
            except Exception as e:
                await message.channel.send(f"Sorry <@{message.author.id}>, I encountered an error: {str(e)}")

@bot.command(name='join')
async def join(ctx):
    """Join the user's voice channel"""
    logger.info("Join command received from %s in guild %s", ctx.author, ctx.guild)
    
    if not ctx.author.voice:
        logger.info("User not in a voice channel")
        await ctx.send("You need to be in a voice channel for me to join!")
        return
        
    if await voice_handler.join_voice_channel(ctx.author.voice.channel):
        logger.info("Successfully joined voice channel %s", ctx.author.voice.channel)
        await ctx.send("👋 I've joined your voice channel! Say 'bot' to get my attention!")
        await voice_handler.start_listening(ctx.guild.id, handle_voice_message)
    else:
        logger.error("Failed to join voice channel")
        await ctx.send("❌ I couldn't join the voice channel!")

@bot.command(name='leave')
async def leave(ctx):
    """Leave the voice channel"""
    logger.info("Leave command received from %s in guild %s", ctx.author, ctx.guild)
    await voice_handler.leave_voice_channel(ctx.guild.id)
    await ctx.send("👋 See ya!")

@bot.command(name='clear')
async def clear(ctx):
    """Clear the conversation history"""
    logger.info("Clear command received from %s in guild %s", ctx.author, ctx.guild)
    conversation_handler.clear_history(ctx.channel.id, guild_id=ctx.guild.id)
    conversation_handler.clear_history(ctx.guild.id, is_voice=True, guild_id=ctx.guild.id)
    await ctx.send("🧹 Conversation history cleared!")

@bot.command(name='ping')
async def ping(ctx):
    """Check bot latency"""
    latency = round(bot.latency * 1000)
    logger.info("Ping command received. Latency: %dms", latency)
    await ctx.send(f'Pong! Latency: {latency}ms')

async def start_bot():
    """Start both the bot and the health check server"""
    # Start health check server
    runner = web.AppRunner(app)
    await runner.setup()
    site = web.TCPSite(runner, '0.0.0.0', 8080)
    await site.start()
    logger.info("Health check server started on port 8080")
    
    # Start the bot
    try:
        await bot.start(TOKEN)
    finally:
        await runner.cleanup()

if __name__ == '__main__':
    # Run both the bot and health check server
    asyncio.run(start_bot()) 